src/environmentd/tests/sql.rs

```diff
@@ -1309,6 +1313,62 @@ fn test_timestamp_recovery() -> Result<(), Box<dyn Error>> {
     Ok(())
 }
 
+#[test]
+fn test_timeline_read_holds() -> Result<(), Box<dyn Error>> {
+    mz_ore::test::init_logging();
+    // Set the timestamp to zero for deterministic initial timestamps.
+    let now = Arc::new(Mutex::new(0));
+    let now_fn = {
+        let now = Arc::clone(&now);
+        NowFn::from(move || *now.lock().unwrap())
+    };
+    let config = util::Config::default().with_now(now_fn).unsafe_mode();
+    let server = util::start_server(config).unwrap();
+    let mut mz_client = server.connect(postgres::NoTls)?;
+
+    let pg_client = create_postgres_source_with_table(
+        &server.runtime,
+        &mut mz_client,
+        "v",
+        "(a INT)",
+        "mz_source",
+    )?;
+
+    // Create user table in Materialize.
+    mz_client.batch_execute(&"DROP TABLE IF EXISTS t;")?;
+    mz_client.batch_execute(&"CREATE TABLE t (a INT);")?;
+    insert_with_deterministic_timestamps("t", "(42)", &server, Arc::clone(&now))?;
+
+    // Insert data into source.
+    let source_rows: i64 = 10;
+    for _ in 0..source_rows {
+        let _ = server
+            .runtime
+            .block_on(pg_client.execute("INSERT INTO v VALUES (42);", &[]))?;
+    }
+
+    // Wait for view to be populated.
+    while mz_client
+        .query_one("SELECT COUNT(*) FROM v;", &[])?
+        .get::<_, i64>(0)
+        != source_rows
+    {
+        thread::sleep(Duration::from_millis(1));
+    }
+
+    // Make sure that the table and view are joinable immediately at some timestamp.
+    let mut mz_join_client = server.connect(postgres::NoTls)?;
+    let _ = mz_ore::test::timeout(Duration::from_millis(1_000), move || {
```

Comment:
The timeout means this test might get flakey. `EXPLAIN TIMESTAMP` is a possible deterministic solution to this because it can answer this question directly. I think this test can be merged as is, but a followup pr can teach EXPLAIN TIMESTAMP to get a new bool field that is "is the selected query timestamp <= all uppers and > all sinces" (I might have a inclusivity error there) maybe named `immediately_available` or something idk.

Reply:
Good idea, what I think I could do for this PR which wouldn't be too hard is to extract the read and write frontiers of every object from the EXPLAIN string. Then I can just check and make sure they all overlap.

Reply:
I recommend against writing a parser for that output. Instead, I recommend teaching EXPLAIN TIMESTAMP to optionally output as JSON then use regular JSON goo to consume it.

Reply:
In that case, I will probably wait for a followup PR.